Svc/FrameAccumulator/FrameAccumulator.cpp

```
@@ -0,0 +1,176 @@
+
+void FrameAccumulator ::dataIn_handler(FwIndexType portNum, Fw::Buffer& buffer, const Drv::RecvStatus& status) {
+    // Check whether there is data to process
+    if (status.e == Drv::RecvStatus::RECV_OK) {
+        // There is: process the data
+        this->processBuffer(buffer);
+    }
+    // Deallocate the buffer
+    this->dataDeallocate_out(0, buffer);
+}
+
+void FrameAccumulator ::processBuffer(Fw::Buffer& buffer) {
+    const U32 bufferSize = buffer.getSize();
+    U8 *const bufferData = buffer.getData();
+    // Current offset into buffer
+    U32 offset = 0;
+    // Remaining data in buffer
+    U32 remaining = bufferSize;
+
+    for (U32 i = 0; i < bufferSize; ++i) {
+        // If there is no data left or no space, exit the loop
+        if (remaining == 0 || this->m_inRing.get_free_size() == 0) {
+            break;
+        }
+        // Compute the size of data to serialize
+        const NATIVE_UINT_TYPE ringFreeSize = this->m_inRing.get_free_size();
+        const NATIVE_UINT_TYPE serSize = (ringFreeSize <= remaining) ?
+                                         ringFreeSize : static_cast<NATIVE_UINT_TYPE>(remaining);
+        // Serialize data into the ring buffer
+        const Fw::SerializeStatus status =
+                this->m_inRing.serialize(&bufferData[offset], serSize);
+        // If data does not fit, there is a coding error
+        FW_ASSERT(
+                status == Fw::FW_SERIALIZE_OK,
+                static_cast<FwAssertArgType>(status),
+                static_cast<FwAssertArgType>(offset),
+                static_cast<FwAssertArgType>(serSize));
+        // Process the data
+        this->processRing();
+        // Update buffer offset and remaining
+        offset += serSize;
+        remaining -= serSize;
+    }
+    // Either all the bytes from the data buffer must be processed, or the ring must be full
+    FW_ASSERT(remaining == 0 || this->m_inRing.get_free_size() == 0, static_cast<FwAssertArgType>(remaining));
```

Does this mean that data will be silently dropped under some circumstances? Can we have a way to make it not silent (either a FATAL or some other kind of response)?